### PR TITLE
mon: handle gw fast-reboot, proper handle of gw delete scenarios

### PR DESCRIPTION
**Pull request by leonidc**
## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/mon/NVMeofGwMap.cc [hunk 1]
```diff
@@ -430,6 +430,62 @@ void  NVMeofGwMap::find_failover_candidate(
   }
 }
 
+void  NVMeofGwMap::handle_gw_performing_fast_reboot(const NvmeGwId &gw_id,
+     const NvmeGroupKey& group_key, bool &map_modified)
+{
+  bool modified = false;
```
Comment (athanatos) [inline src/mon/NVMeofGwMap.cc:436]: Why not just pass map_modified to fsm_handle_gw_fast_reboot?

File: src/mon/NVMeofGwMap.cc [hunk 2]
```diff
@@ -430,6 +430,62 @@ void  NVMeofGwMap::find_failover_candidate(
...
+  for (auto& state_itr: created_gws[group_key][gw_id].sm_state ) {
+    fsm_handle_gw_fast_reboot(gw_id,group_key, state_itr.first, modified);
+    map_modified |= modified;
+  }
+}
+
+void NVMeofGwMap::fsm_handle_gw_fast_reboot(const NvmeGwId &gw_id,
+    const NvmeGroupKey& group_key, NvmeAnaGrpId grpid, bool &map_modified) {
```
Comment (athanatos) [inline src/mon/NVMeofGwMap.cc:444]: Brace on new line for definitions in .cc files as you did in handle_gw_performing_fast_reboot above.

**Review by athanatos — APPROVED** [review]
Commit is missing a signed-off by -- required on all commits.  LGTM once the commit message and the below changes are made.

**Comment by athanatos** [comment]
Pushed testing branch https://shaman.ceph.com/builds/ceph/wip-sjust-nvmeof-testing-2024-08-21 with https://github.com/ceph/ceph/pull/59240, https://github.com/ceph/ceph/pull/59385, https://github.com/ceph/ceph/pull/59388, and https://github.com/ceph/ceph/pull/59366

**Comment by caroav** [comment]
jenkins test api

**Comment by caroav** [comment]
jenkins test make check

File: src/mon/NVMeofGwMap.cc (hunk 1 above)
Comment (leonidc) [reply src/mon/NVMeofGwMap.cc:436]: sure

**Comment by leonidc** [comment]
jenkins test api

File: NVMeofGwMap.h
Comment (cbodley) [inline NVMeofGwMap.h:1]: this added a duplicate copy of ceph/src/mon/NVMeofGwMap.h at ceph/NVMeofGwMap.h. i opened https://github.com/ceph/ceph/pull/59840 to remove this

**Pull request merged**